### DISCUSSION TITLE
How to use Randomizable third party transforms or customered transforms for CacheDataset?

### DISCUSSION THREAD

**Discussion by Jingnan-Jia** [Project-MONAI/MONAI#2796] [Q&A] [answered]
[This jupyter](https://github.com/Project-MONAI/tutorials/blob/master/modules/integrate_3rd_party_transforms.ipynb) shows how to insert third party transforms (from `torchIO` and `rising`) into `monai`. But all of the transforms are not `Randomizable`, and it used the normal `Dataset` instead of `CacheDataset` or `SmartCacheDataset`. 

I noted that The `CacheDataset` or `SmartCacheDataset` will cache the fixed transforms at first, leave the `Randomizable` transforms to be implemented by `Dataloader`. 

The `Randomizable` transforms from `Monai` [inherite `Randomizable` class](https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/spatial/dictionary.py#:~:text=class%20RandRotate90d(RandomizableTransform%2C%20MapTransform%2C%20InvertibleTransform)%3A) which would be [cached in this line.](https://github.com/Project-MONAI/MONAI/blob/6e025069bfc15ee325661601873f08f751695cda/monai/data/dataset.py#:~:text=the%20deterministic%20transforms-,if%20isinstance(_transform%2C%20Randomizable)%20or%20not%20isinstance(_transform%2C%20Transform)%3A,break,-%23%20this%20is%20to)

It means that only the transforms which inherite `Randomizable` and `Transforms` would be possible to be cached. 

However our customed transforms or the transforms from torchIO does not inherite `Randomizable` and `Transforms`, which means our customed transforms and the torchIO.transforms can not be cached by `CacheDataset` or `SmartCacheDataset`.

For instance, this is my own customed transform to load data (I **hope** it could be cached):

```
class LoadDatad:
    def __call__(self, data: Mapping[str, Union[np.ndarray, str]]) -> Dict[str, np.ndarray]:
        path = data['path_key']
        x = load_itk(fpath, require_ori_sp=True)

        data = {'image_key': x,
                    'path_key': path}  
        return data
```

If I use this transform as the first transform in my composed transform list, I found this transform will not be cached at all. 

So I think it is necessary to remind users in the documentation that "If you want your own transforms be cached by monai.CacheDataset/SmartCacheDataset/PersistentDataset, you have to ensure: 1. your transforms inherite monai.Transform 2. your transforms can not be Randomizable "

Another solution is: changing [this line](https://github.com/Project-MONAI/MONAI/blob/6e025069bfc15ee325661601873f08f751695cda/monai/data/dataset.py#:~:text=the%20deterministic%20transforms-,if%20isinstance(_transform%2C%20Randomizable)%20or%20not%20isinstance(_transform%2C%20Transform)%3A,break,-%23%20this%20is%20to): 
from
`if isinstance(_transform, Randomizable) or not isinstance(_transform, Transform):`
to 
`if isinstance(_transform, Randomizable):`

Is my understanding right?

**Comment by Nic-Ma** [comment, marked as answer]
Hi @Jingnan-Jia ,

Thanks for your investigation and analysis & suggestions.
I submitted PR to enhance the doc-string of cache datasets: https://github.com/Project-MONAI/MONAI/pull/2799.

Thanks.
